### PR TITLE
MINIFICPP-1440 PublishKafka processor's "security protocol" should ha…

### PR DESCRIPTION
**Pull request by dam4rus**
…ve allowable values

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/librdkafka/PublishKafka.cpp [hunk 1]
```diff
@@ -126,7 +126,11 @@ const core::Property PublishKafka::CompressCodec(
 const core::Property PublishKafka::MaxFlowSegSize(
     core::PropertyBuilder::createProperty("Max Flow Segment Size")->withDescription("Maximum flow content payload segment size for the kafka record. 0 B means unlimited.")
         ->isRequired(false)->withDefaultValue<core::DataSizeValue>("0 B")->build());
-const core::Property PublishKafka::SecurityProtocol("Security Protocol", "Protocol used to communicate with brokers", "");
+const core::Property PublishKafka::SecurityProtocol(
+        core::PropertyBuilder::createProperty("Security Protocol")
+        ->withDescription("Protocol used to communicate with brokers")
+        ->withAllowableValues<std::string>({SECURITY_PROTOCOL_PLAINTEXT, SECURITY_PROTOCOL_SSL, SECURITY_PROTOCOL_SASL_PLAINTEXT, SECURITY_PROTOCOL_SASL_SSL})
```
Comment (szaszm) [inline extensions/librdkafka/PublishKafka.cpp:132]: When looking at `PublishKafka.cpp:686`, it seems like only `SECURITY_PROTOCOL_SSL` is actually supported, and the default empty value is interpreted as plaintext.

I suggest restricting the set of "allowable values" to plaintext and ssl, and handling plaintext near `PublishKafka.cpp:687`. I think doing nothing is a valid handling of plaintext.

**Review by szaszm — COMMENTED** [review]
Thanks for your first contribution. We generally merge PRs after 2-3 approvals and when all comments are resolved. In case there are any questions, feel free to reach out.

File: extensions/librdkafka/PublishKafka.cpp (hunk 1 above)
Comment (hunyadi-dev) [reply extensions/librdkafka/PublishKafka.cpp:132]: I would also add a default value for `PLAINTEXT` as we are trying to make the behaviour consistent with NiFi:
https://nifi.apache.org/docs/nifi-docs/components/nifi-docs/components/org.apache.nifi/nifi-kafka-2-0-nar/1.9.0/org.apache.nifi.processors.kafka.pubsub.PublishKafka_2_0/

Comment (dam4rus) [reply extensions/librdkafka/PublishKafka.cpp:132]: > When looking at `PublishKafka.cpp:686`, it seems like only `SECURITY_PROTOCOL_SSL` is actually supported, and the default empty value is interpreted as plaintext.

Should I leave the unused defines in? If it doesn't breaks backward compatibility removing them would make the code a bit clearer.

> I suggest restricting the set of "allowable values" to plaintext and ssl, and handling plaintext near PublishKafka.cpp:687. I think doing nothing is a valid handling of plaintext.

What's the suggested coding style in this case? Having an empty `else if (value == SECURITY_PROTOCOL_PLAINTEXT)` branch or just using `else if (value != SECURITY_PROTOCOL_PLAINTEXT)` to handle unexpected values?

> I would also add a default value for PLAINTEXT as we are trying to make the behaviour consistent with NiFi:
https://nifi.apache.org/docs/nifi-docs/components/nifi-docs/components/org.apache.nifi/nifi-kafka-2-0-nar/1.9.0/org.apache.nifi.processors.kafka.pubsub.PublishKafka_2_0/

Nifi also describes it as a required value. Should I add that too?

Comment (hunyadi-dev) [reply extensions/librdkafka/PublishKafka.cpp:132]: 1. Feel free to delete unused parts.
2. Up to you. If you feel fancy, you can extend the property API to be able to return the allowable values and check against presence there.
3. Yes.  The more important thing is that we are compatible with NiFi configs, but compatibility in the other direction is probably nice to have.

**Review by szaszm — APPROVED** [review]

**Review by fgerlits — APPROVED** [review]

**Review by hunyadi-dev — APPROVED** [review]

**Pull request closed without merging**